heartbeat/monitors/active/http/task.go

```diff
@@ -260,12 +261,30 @@ func execRequest(client *http.Client, req *http.Request, validator func(*http.Re
 	}
 
 	err = validator(resp)
-	end = time.Now()
 	if err != nil {
-		return start, end, resp, reason.ValidateFailed(err)
+		return start, time.Now(), resp, reason.ValidateFailed(err)
+	}
+
+	// Read the entirety of the body. Otherwise, the stats for the check
+	// don't include download time.
+	buf := make([]byte, 1024)
+	for {
+		read, err := resp.Body.Read(buf)
+		if read < 0 {
+			panic("negative read encountered in http response!")
```

Comment:
Do you really want to stop the application if this happens? I don't see a reason why it should happen but still not sure if it would be helpful to exit the application? I would rather print out an error message to be able to debug it and everything else keeps running.

Reply:
Yes, this is what the stdlib does. I shamelessly copied the behavior from: https://github.com/golang/go/blob/master/src/bytes/buffer.go#L202

Reply:
I was curious about why we check for that, I think its more of a defensive check against bad implementation of `io.Reader`. Looking at the method signature we do return an _int_ and not an _uint_. But returning a negative number would make everything fails, I haven't personnaly added an assert to a Read return valus but I do not think it hurts. :)

```
type Reader interface {
	Read(p []byte) (n int, err error)
}
```

Reply:
What about using [io.Copy](https://golang.org/pkg/io/#Copy)? or [io.ReadFull](https://golang.org/pkg/io/#ReadFull)?

Reply:
@ph I didn't use those because for the case of large bodies I didn't want to have to read everything into memory. In the forum post that helped me find this issue https://discuss.elastic.co/t/using-heartbeat-as-a-speed-test/155033/3 the user was using heartbeat to check very large bodies. In that case we should do the equivalent of a read to dev null.

Reply:
I've now simplified this with a `Discard` writer.